File: exercise/C#/day24/src/Delivery/UseCases/ToyDeliveryUseCase.cs
using Delivery.Domain;
using Delivery.Domain.Core;
using LanguageExt;
using static Delivery.Domain.Core.Error;

namespace Delivery.UseCases
{
    public class ToyDeliveryUseCase(IToyRepository repository)
    {
        public Either<Error, Unit> Handle(DeliverToy deliverToy)
            => repository
                .FindByName(deliverToy.DesiredToy)
                .ToEither(() => ErrorFor(deliverToy))
                .Bind(ReduceStock)
                .Map(_ => Unit.Default);

        private Either<Error, Toy> ReduceStock(Toy toy)
            => toy.ReduceStock()
                .Let(_ => repository.Save(toy));

        private static Error ErrorFor(DeliverToy deliverToy)
            => AnError($"Oops we have a problem... we have not build the toy: {deliverToy.DesiredToy}");
    }
}

File: exercise/php/day24/src/Delivery/UseCases/ToyDeliveryUseCase.php
<?php

namespace Delivery\UseCases;

use Delivery\Domain\IToyRepository;

class ToyDeliveryUseCase
{
    private IToyRepository $repository;

    public function __construct(IToyRepository $repository)
    {
        $this->repository = $repository;
    }

    /**
     * @throws \Exception
     */
    public function handle(DeliverToy $deliverToy): void
    {
        $foundToy = $this->repository->findByName(
            $deliverToy->getDesiredToy()
        );

        if (!$foundToy) {
            throw new \Exception("Oops we have a problem... we have not built the toy: {$deliverToy->getDesiredToy()}");
        }

        $foundToy->reduceStock();
        $this->repository->save($foundToy);
    }
}